### ISSUE TITLE
interface extending a type produces error only if a generic is changed

### ISSUE THREAD

**Issue by ackvf** [microsoft/TypeScript#24415]
**TypeScript Version:**  3.0.0-dev.20180522

**Code**
This code does not produce error.
```ts
type CurrencyObject<T, K extends string> = {
  [P in K]: T
}

interface InnerType {
  unit: string,
  decimals: number,
  prefix: string,
  postfix: string,
}

interface SYMBOLSType<T, K> extends CurrencyObject<T, MySymbols> { // <- notice "MySymbols" here
  format: (value: number, currency: K) => string
}

type MySymbols = 'BTC' | 'CZK' | 'ETH' | 'EUR' | 'LTC'

const SYMBOLS: SYMBOLSType<InnerType, MySymbols> = {
  BTC: { unit: 'BTC', decimals: 5, prefix: '',  postfix: ''   },
  CZK: { unit: 'CZK', decimals: 2, prefix: '',  postfix: ',-' },
  ETH: { unit: 'ETH', decimals: 4, prefix: 'Ξ', postfix: ''   },
  EUR: { unit: 'EUR', decimals: 2, prefix: '€', postfix: ''   },
  LTC: { unit: 'LTC', decimals: 3, prefix: '',  postfix: ''   },
  format(value, currency) {
    const c = this[currency]
    return c.prefix + value.toFixed(c.decimals) + c.postfix
  },
}
```

Changing `MySymbols` to `K` starts to give an error: `An interface may only extend a class or another interface.`

```ts
interface SYMBOLSType<T, K extends string> extends CurrencyObject<T, K> { // error
  format: (value: number, currency: K) => string
}
```

**Expected behavior:**
Expected is that I am able to reuse the generic `K` so that I can remove the `type MySymbols` and use the union inline `const SYMBOLS: SYMBOLSType<InnerType, 'BTC' | 'CZK' | 'ETH' | 'EUR' | 'LTC'> = {`

**Actual behavior:**
I am not able to reuse the generic `K` with a seemingly incorrect error as the exact code above does not produce the error.

But if the error is right and I cannot extend a `type`, the code above should not work. (but it does)

**Comment by mhegazy** [comment]
The compiler needs to know the shape of the type in an extends clause. the rational here is that `extends` for interfaces and classes have some additional checks for sustainability, and these can not be done correctly unless the shape of the type is known. for mapped types on generic types, that is not known until instantiation time.. that is like saying `interface I<T>  extends T { }`, and the compiler can not verify that by looking at the declaration, so it is disallowed.

you can use intersection types instead to achieve a similar result., e.g. `type SYMBOLSType<T, K> = I & CurrencyObject<T, K>`.

**Comment by ackvf** [comment]
Thank you for clarification and a lesson.
I did use the intersection in my first snippet, but I thought interfaces are a more clean approach. After all, the linter does tell me to use interfaces instead of types and since I used them I thought about using their features as well.

Anway, since the error message is confusing right now, would it be possible to slightly modify it?

**Comment by mhegazy** [comment]
Open for proposals for a better error message.

**Comment by ackvf** [comment]
Can you put another label on it too, to mark it for future input?

**Comment by RyanCavanaugh** [comment]
All things are open to future inputs. The labels here are what we'd like them to be.

**Comment by ackvf** [comment]
I meant no offense @RyanCavanaugh, but clearly if it is labeled as "Working as Intended", I feel that it will be burried in the history.

**Comment by RyanCavanaugh** [comment]
This hasn't garnered any additional feedback in several years, and a way of declaring the type exists. I don't think we need to make further investment in this particular scenario.
